narwhals/functions.py

```diff
@@ -1775,3 +1779,148 @@ def coalesce(
+        └───────────────────────────┘
+
+        `end` can be omitted for a shorter syntax.
+
+        >>> nw.int_range(5, step=2, eager="pandas")
+        ┌───────────────────────────┐
+        |      Narwhals Series      |
+        |---------------------------|
+        |0    0                     |
+        |1    2                     |
+        |2    4                     |
+        |Name: literal, dtype: int64|
+        └───────────────────────────┘
+
+        Generate an index column by using `int_range` in conjunction with :func:`len`.
+
+        >>> import pandas as pd
+        >>> df = nw.from_native(pd.DataFrame({"a": [1, 3, 5], "b": [2, 4, 6]}))
+        >>> df.select(nw.int_range(nw.len(), dtype=nw.UInt32).alias("index"), nw.all())
+        ┌──────────────────┐
+        |Narwhals DataFrame|
+        |------------------|
+        |     index  a  b  |
+        |  0      0  1  2  |
+        |  1      1  3  4  |
+        |  2      2  5  6  |
+        └──────────────────┘
+    """
+    return _int_range_impl(start, end, step, dtype=dtype, eager=eager)
+
+
+def _int_range_impl(
+    start: int | Expr,
+    end: int | Expr | None,
+    step: int,
+    *,
+    dtype: IntegerDType,
+    eager: ModuleType | Implementation | str | Literal[False],
+) -> Expr | Series[Any]:
+    from narwhals.exceptions import ComputeError
+
+    if not dtype.is_integer():
+        msg = f"non-integer `dtype` passed to `int_range`: {dtype}"
+        raise ComputeError(msg)
+
+    if end is None:
+        end = start
+        start = 0
+
+    if not eager:
+        assert start is not None  # noqa: S101, help mypy
+        assert end is not None  # noqa: S101, help mypy
+
+        start = start if isinstance(start, Expr) else lit(start, dtype=dtype)
+        end = end if isinstance(end, Expr) else lit(end, dtype=dtype)
+
+        if start._metadata.expansion_kind.is_multi_output():
+            msg = "`start` must contain exactly one value, got expression returning multiple values"
+            raise ComputeError(msg)
+
+        if end._metadata.expansion_kind.is_multi_output():
+            msg = "`end` must contain exactly one value, got expression returning multiple values"
+            raise ComputeError(msg)
+
+        return Expr(
+            lambda plx: apply_n_ary_operation(
+                plx,
+                lambda *args: plx.int_range(*args, dtype=dtype),
+                start,
+                end,
+                step,
+                str_as_lit=False,
+            ),
+            ExprMetadata.selector_single(),
+        )
+
+    impl = Implementation.from_backend(eager)
+    if is_eager_allowed(impl):
+        assert isinstance(start, int)  # noqa: S101, help mypy
+        assert isinstance(end, int)  # noqa: S101, help mypy
+        ns = Version.MAIN.namespace.from_backend(impl).compliant
+        series = ns._series._int_range(
+            start=start, end=end, step=step, dtype=dtype, context=ns, name="literal"
+        )
```

Comment:
Probably my most common review comment is to beg that we don't use `assert` for typing.

Currently, this is hiding the fact that we aren't handling passing expressions **and** `eager`

```suggestion
    if is_eager_allowed(impl):
        ns = Version.MAIN.namespace.from_backend(impl).compliant
        series = ns._series._int_range(
            start=start, end=end, step=step, dtype=dtype, context=ns, name="literal"
        )
```

I've added a test with two options for solving the issue.

I'm 100% fine with not supporting `Expr` for `Series`, but would like a helpful error and typing to match this difference from `polars` 🙏

Reply:
Those tests currently raise here

```py
        impl = Implementation.from_backend(eager)
        if is_eager_allowed(impl):
>           assert isinstance(start, int)  # noqa: S101, help mypy
                   ^^^^^^^^^^^^^^^^^^^^^^
E           AssertionError

narwhals/functions.py:1914: AssertionError
==== short test summary info ===
FAILED tests/expr_and_series/int_range_test.py::test_int_range_eager_expr[polars[eager]] - AssertionError
FAILED tests/expr_and_series/int_range_test.py::test_int_range_eager_expr[pandas] - AssertionError
FAILED tests/expr_and_series/int_range_test.py::test_int_range_eager_expr[pandas[pyarrow]] - AssertionError
FAILED tests/expr_and_series/int_range_test.py::test_int_range_eager_expr[pyarrow] - AssertionError
```

Reply:
@dangotbanned I am addressing all the comments.

I don't think it is allowed in polars:

```py
pl.int_range(pl.col("a").min(), eager=True)
```
> ColumnNotFoundError: a

Reply:
Oof 😅

In that case, just this part then

> I'm 100% fine with not supporting `Expr` for `Series`, but would like a helpful error and typing

And **(1)** from https://github.com/narwhals-dev/narwhals/pull/2895/commits/4706a30f8652c143ed42a208f259f18a6b1246a9